src/gentropy/datasource/ukb_ppp_eur/summary_stats.py

```diff
@@ -0,0 +1,162 @@
+
+from dataclasses import dataclass
+
+import pyspark.sql.functions as f
+import pyspark.sql.types as t
+from pyspark.sql import SparkSession
+
+from gentropy.common.spark_helpers import neglog_pvalue_to_mantissa_and_exponent
+from gentropy.dataset.summary_statistics import SummaryStatistics
+
+
+@dataclass
+class UkbPppEurSummaryStats:
+    """Summary statistics dataset for UKB PPP (EUR)."""
+
+    @classmethod
+    def from_source(
+        cls: type[UkbPppEurSummaryStats],
+        spark: SparkSession,
+        raw_summary_stats_path: str,
+        tmp_variant_annotation_path: str,
+        chromosome: str,
+    ) -> SummaryStatistics:
+        """Ingests all summary stats for UKB PPP (EUR) data.
+
+        Args:
+            spark (SparkSession): Spark session object.
+            raw_summary_stats_path (str): Input raw summary stats path.
+            tmp_variant_annotation_path (str): Input variant annotation dataset path.
+            chromosome (str): Which chromosome to process.
+
+        Returns:
+            SummaryStatistics: Processed summary statistics dataset for a given chromosome.
+        """
+        # Read the precomputed variant annotation dataset.
+        va_df = (
+            spark
+            .read
+            .parquet(tmp_variant_annotation_path)
+            .filter(f.col("vaChromosome") == ("X" if chromosome == "23" else chromosome))
+            .persist()
+        )
+
+        # Read and process the summary stats dataset.
+        df = (
+            spark
+            .read
+            .parquet(raw_summary_stats_path)
+            .filter(f.col("chromosome") == chromosome)
+            # Harmonise, 1: Rename chromosome 23 to X.
+            .withColumn(
+                "chromosome",
+                f.when(
+                    f.col("chromosome") == "23", "X"
+                ).otherwise(f.col("chromosome"))
+            )
+            # Harmonise, 2: Filter out low INFO rows.
+            .filter(f.col("INFO") >= 0.8)
+            # Harmonise, 3: Filter out low frequency rows.
+            .withColumn(
+                "freq",
+                f.when(f.col("A1FREQ") < 0.5, f.col("A1FREQ"))
+                .otherwise(1 - f.col("A1FREQ"))
+            )
+            .filter(f.col("freq") >= 0.0001)
+            .drop("freq")
+            # Harmonise, 4: Assign variant types.
+            .withColumn(
+                "variant_type",
+                f.when(
+                    (f.length("ALLELE0") == 1) & (f.length("ALLELE1") == 1),
+                    f.when(
+                        ((f.col("ALLELE0") == "A") & (f.col("ALLELE1") == "T")) |
+                        ((f.col("ALLELE0") == "T") & (f.col("ALLELE1") == "A")) |
+                        ((f.col("ALLELE0") == "G") & (f.col("ALLELE1") == "C")) |
+                        ((f.col("ALLELE0") == "C") & (f.col("ALLELE1") == "G")),
+                        "snp_c"
+                    )
+                    .otherwise(
+                        "snp_n"
```

Comment:
What is the meaning behind `snp_c` and `snp_n` ?

Reply:
This notation comes from the investigations in the linked issue:

> Variants are also split into three types:
>
> * snp_c: SNP which replaces a nucleotide with its complement, so four types: A → T, T → A, G → C, C → G
> * snp_n: All other SNPs
> * indel: Indels, defined as variants where the length of at least one allele is > 1.

https://github.com/opentargets/issues/issues/3234#issuecomment-2069173983